File: src/types/scheduler.ts
export type SchedulerTaskStatus =
  | "idle"
  | "running"
  | "succeeded"
  | "failed"
  | "paused"
  | "completed";

export type SchedulerRunResult = "success" | "failed";

export type SchedulerTrigger =
  | { type: "once"; runAt: string }
  | { type: "daily"; time: string }
  | { type: "weekly"; weekdays: number[]; time: string }
  | { type: "countdown"; durationSeconds: number };

export interface SchedulerAction {
  title: string;
  message: string;
  enableSystemNotification: boolean;
  enableInAppAlert: boolean;
  enableSound: boolean;
}

export interface SchedulerTask {
  id: string;
  name: string;
  description?: string | null;
  enabled: boolean;
  status: SchedulerTaskStatus;
  trigger: SchedulerTrigger;
  action: SchedulerAction;
  createdAt: string;
  updatedAt: string;
  lastRunAt?: string | null;
  nextRunAt?: string | null;
}

export interface SchedulerTaskInput {
  name: string;
  description?: string | null;
  enabled: boolean;
  trigger: SchedulerTrigger;
  action: SchedulerAction;
}

export interface SchedulerTaskUpdate extends SchedulerTaskInput {
  id: string;
}

export interface SchedulerTaskLog {
  id: string;
  taskId: string;
  taskName: string;
  scheduledAt?: string | null;
  executedAt: string;
  result: SchedulerRunResult;
  errorMessage?: string | null;
  durationMs: number;
}

export interface ReminderEventPayload {
  taskId: string;
  taskName: string;
  title: string;
  message: string;
  firedAt: string;
  showInAppAlert: boolean;
  playSound: boolean;
}

File: src-tauri/src/models/scheduler.rs
//! 定时中心数据模型

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SchedulerTask {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    pub enabled: bool,
    pub status: SchedulerTaskStatus,
    pub trigger: SchedulerTrigger,
    pub action: SchedulerAction,
    pub created_at: String,
    pub updated_at: String,
    pub last_run_at: Option<String>,
    pub next_run_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SchedulerTaskInput {
    pub name: String,
    pub description: Option<String>,
    pub enabled: bool,
    pub trigger: SchedulerTrigger,
    pub action: SchedulerAction,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SchedulerTaskUpdate {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    pub enabled: bool,
    pub trigger: SchedulerTrigger,
    pub action: SchedulerAction,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SchedulerTaskLog {
    pub id: String,
    pub task_id: String,
    pub task_name: String,
    pub scheduled_at: Option<String>,
    pub executed_at: String,
    pub result: SchedulerRunResult,
    pub error_message: Option<String>,
    pub duration_ms: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReminderEventPayload {
    pub task_id: String,
    pub task_name: String,
    pub title: String,
    pub message: String,
    pub fired_at: String,
    pub show_in_app_alert: bool,
    pub play_sound: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum SchedulerTaskStatus {
    Idle,
    Running,
    Succeeded,
    Failed,
    Paused,
    Completed,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum SchedulerRunResult {
    Success,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(
    tag = "type",
    rename_all = "camelCase",
    rename_all_fields = "camelCase"
)]
pub enum SchedulerTrigger {
    Once { run_at: String },
    Daily { time: String },
    Weekly { weekdays: Vec<u8>, time: String },
    Countdown { duration_seconds: i64 },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SchedulerAction {
    pub title: String,
    pub message: String,
    pub enable_system_notification: bool,
    pub enable_in_app_alert: bool,
    pub enable_sound: bool,
}
